File: January/Day007/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 * int val;
 * TreeNode *left;
 * TreeNode *right;
 * TreeNode() : val(0), left(nullptr), right(nullptr) {}
 * TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // Vector to store the sum of every subtree
    vector<long long> sums;

    // Helper function to compute subtree sums
    long long calculateSubtreeSum(TreeNode* node) {
        if (node == nullptr) return 0;

        // Recursively get left and right sums
        long long leftSum = calculateSubtreeSum(node->left);
        long long rightSum = calculateSubtreeSum(node->right);

        // Current subtree sum = node value + left + right
        long long currentSum = node->val + leftSum + rightSum;

        // Store the sum in our vector
        sums.push_back(currentSum);

        return currentSum;
    }

    int maxProduct(TreeNode* root) {
        sums.clear(); // Clear vector for safety

        // 1. Get total sum of the tree
        long long totalSum = calculateSubtreeSum(root);
        long long maxProd = 0;

        // 2. Iterate through all sums to find the maximum product
        // We use 'long long' because the product can exceed the 32-bit integer limit
        for (long long s : sums) {
            long long currentProd = s * (totalSum - s);
            maxProd = max(maxProd, currentProd);
        }

        // 3. Return result modulo 10^9 + 7
        return (int)(maxProd % 1000000007);
    }
};

File: January/Day007/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        # List to store the sum of every subtree encountered
        all_sums = []

        # Helper function to calculate sums via Post-Order Traversal
        def subtree_sum(node):
            if not node:
                return 0
            
            # Recursively get sum of left and right subtrees
            left_sum = subtree_sum(node.left)
            right_sum = subtree_sum(node.right)
            
            # Current subtree sum = node value + left sum + right sum
            current_sum = node.val + left_sum + right_sum
            
            # Add to our list of sums
            all_sums.append(current_sum)
            
            return current_sum

        # 1. Calculate the total sum of the entire tree
        # The last value computed by the recursive function on 'root' is the total sum
        total_sum = subtree_sum(root)
        
        maximum_product = 0
        
        # 2. Iterate through all calculated subtree sums to find the max product
        for s in all_sums:
            # If we cut the edge above this subtree, the two parts are:
            # 's' and 'total_sum - s'
            product = s * (total_sum - s)
            maximum_product = max(maximum_product, product)
            
        # 3. Return result modulo 10^9 + 7
        return maximum_product % (10**9 + 7)